loader.js

```diff
@@ -1276,6 +1385,103 @@ var Module = null;
      getfullscreenenabler().call(this._canvas);
    };
 
+   /*
+    * V86Runner
+    */
+   function V86Runner(canvas, game_data) {
+     // v86 needs a specific DOM structure instead of a canvas
+     var screenContainerOuterElt = document.createElement("div");
+     screenContainerOuterElt.id = canvas.id;
+     screenContainerOuterElt.classList = canvas.classList;
+     screenContainerOuterElt.style = canvas.style;
+
+     var screenContainerInnerElt = document.createElement("div");
+     screenContainerInnerElt.classList = ["emularity-v86-screen-container"];
+     screenContainerInnerElt.style = "display:flex;justify-content:center;align-items:center;background-color:#000;";
+
+     var textDivElt = document.createElement("div");
+     textDivElt.classList = ["emularity-v86-screen-text"];
+     textDivElt.style = "font-size:14px;font-family:monospace;line-height:14px;white-space:pre;";
+     var canvasElt = document.createElement("canvas");
+     canvasElt.classList = ["emularity-v86-screen-canvas"];
+     canvasElt.style = "display:none;";
+
+     screenContainerInnerElt.appendChild(textDivElt);
+     screenContainerInnerElt.appendChild(canvasElt);
+     screenContainerOuterElt.appendChild(screenContainerInnerElt);
+     canvas.parentNode.replaceChild(screenContainerOuterElt, canvas);
+
+     var cfg = {};
+     cfg.screen_container = screenContainerInnerElt;
+     cfg.memory_size = game_data.memory_size || 32 << 20;
+     cfg.vga_memory_size = game_data.vga_memory_size || 2 << 20;
+     cfg.acpi = game_data.acpi;
+     cfg.boot_order = game_data.boot_order || 0x213;
+     cfg.autostart = true;
+
+     cfg.wasm_fn = env => {
+       return new Promise(async resolve => {
+         const wasm = await WebAssembly.instantiate(game_data.wasmBinary, env);
+         resolve(wasm.instance.exports);
+       });
+     };
+     ["bios", "vga_bios", "fda", "fdb", "cdrom", "hda", "hdb"].forEach(key => {
+       if (game_data[key] && game_data[key]["path"]) {
+         cfg[key] = {"buffer": game_data.fs.readFileSync('/'+game_data[key]["path"], null, flag_r).buffer};
```

Comment:
Possibly the same with these as well, using a helper function to reduce the duplication. It’s up to you though.

Reply:
Wouldn't that be too early, though? Since the order is Loader gets instantiated -> Loader gets configured -> files get downloaded -> Runner gets instantiated; as such, my assumption was that I can't just pass in the completed ArrayBuffers in the Loader's configuration stage.

Reply:
I’m glad one of us is paying attention :)